framework/db/QueryBuilder.php

```diff
@@ -305,6 +315,115 @@ public function batchInsert($table, $columns, $rows)
         . ' (' . implode(', ', $columns) . ') VALUES ' . implode(', ', $values);
     }
 
+    /**
+     * Creates an SQL statement to insert rows into a database table if
+     * they do not already exist (matching unique constraints),
+     * or update them if they do.
+     *
+     * For example,
+     *
+     * ```php
+     * $sql = $queryBuilder->upsert('pages', [
+     *     'name' => 'Front page',
+     *     'url' => 'http://example.com/', // url is unique
+     *     'visits' => 0,
+     * ], [
+     *     'visits' => new \yii\db\Expression('visits + 1'),
+     * ], $params);
+     * ```
+     *
+     * The method will properly escape the table and column names.
+     *
+     * @param string $table the table that new rows will be inserted into/updated in.
+     * @param array|Query $insertColumns the column data (name => value) to be inserted into the table or instance
+     * of [[Query]] to perform `INSERT INTO ... SELECT` SQL statement.
+     * @param array|bool $updateColumns the column data (name => value) to be updated if they already exist.
+     * If `true` is passed, the column data will be updated to match the insert column data.
+     * If `false` is passed, no update will be performed if the column data already exists.
```

Comment:
Use case: 

I have a table with 20 columns. On upsert, if record exists, I want to update all column except `created_at`. How to do this with this signature? I need to repeat all columns (except one) with own expressions?

Reply:
Then you have to pass an array of `Expression`s with `VALUES(column)`/`EXCLUDED.column` depending on a DBMS.
I forgot to note this syntax, good catch. :)

Reply:
Honestly, a don't really like this syntax. It took me like 20 minutes to get how this works and results were really surprised to me. I would expect that `$updateColumns` would work similar to `select()`:

1. `['first_name', 'last_name']` will update only `first_name` and `last_name` columns:
   ```ON DUPLICATE KEY UPDATE `first_name`=VALUES(`first_name`), `last_name`=VALUES(`last_name`)```.
2. `['first_name', 'old_first_name' => new Expression('first_name')]` will update `first_name` column and put previous value of `first_name` into `old_first_name`:
   ```ON DUPLICATE KEY UPDATE `first_name`=VALUES(`first_name`), `old_first_name`=first_name```.
3. `['new_first_name' => 'first_name']` will update `new_first_name`column with new `first_name` value:
   ```ON DUPLICATE KEY UPDATE `new_first_name`=VALUES(`first_name`)```.
4. `['first_name' => new Expression('"rob006"')]` will always set `rob006` as `first_name` if record exist:
    ```ON DUPLICATE KEY UPDATE `first_name`="rob006"```.

This should simplify every non-standard (when you don't need just update all columns with inserted values) case and can be used in the same way in `upsertMultiple()`.

Reply:
TBH, I really like this idea. As you could see from the current implementation, it will require a "bit" of magic but it's definitely possible.
I'm going to try to implement it as soon as everything described in https://github.com/yiisoft/yii2/pull/15276#issuecomment-350549841 land.